File: Julia/combinatorial.jl
# combinatorial.jl: Combinatorial functions.

# (c) E. Adrian Garro S. Costa Rica Institute of Technology.

"""
`combinations_of_2(xs)`
Combinations of 2 without repeated elements.
"""
function combinations_of_2(xs)
	return [(x, y) for x in xs for y in xs if y > x]
end
	
"""
`permutations(xs)`
Permutations without repeated elements.
"""
function permutations(xs)
	if xs == []
		produce(Int[])
	else
		for x in xs
			temp = copy(xs)
			filter!(e -> e != x, temp)
			for ys in @task permutations(temp)
				produce(union(Int[x], ys))
			end
		end
	end
end

File: Python/combinatorial.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

'''
combinatorial.py: Combinatorial functions.
'''

__copyright__ = '(c) 2016 E. Adrian Garro S. Costa Rica Institute of Technology.'

def combinations_of_2(xs):
    '''
    Combinations of 2 without repeated elements.
    '''
    return ((x, y) for x in xs for y in xs if y > x)

def permutations(xs):
    '''
    Permutations without repeated elements.
    '''
    if xs == []:
        yield []
    for x in xs:
        temp = xs[:]
        temp.remove(x)
        for ys in permutations(temp):
            yield [x] + ys
